[John Ng] Duke Increments

JohnNzj opened the pull request.

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

j-lum committed:
Change file mode on `gradle` to be executable

j-lum committed:
Merge pull request #12 from j-lum/gradle+x

Change file mode on `gradle` to be executable (#9)

JohnNzj committed:
Add Echo, Exit

JohnNzj committed:
Add, List

JohnNzj committed:
Add Task.java, Mark tasks as Done

JohnNzj committed:
Add Todo.java, Event.java, Deadline.java,

JohnNzj committed:
Add DukeException.java, Modified Duke.java to handle Errors

JohnNzj committed:
Add delete

JohnNzj committed:
Add data/duke.txt,Duke.java to save, load from file

JohnNzj committed:
Duke.java: modify to understand date in the form of dd/mm/yyyy hhmm

JohnNzj committed:
Merge branch 'branch-Level-7'

JohnNzj committed:
Merge branch 'branch-Level-8'
duke.txt: add more events, deadlines
# Conflicts:
#	src/main/java/Duke.java

JohnNzj committed:
Add Ui.java, TaskList.java, Storage.java, Parser.java, Command.java.
Modify Duke.java & duke.txt

JohnNzj committed:
Add ParserTest.java & TaskListTest.java to test Parser and TaskList

JohnNzj committed:
Fixed bugs regarding to exception handling for empty task name

JohnNzj committed:
Add JavaDoc comments to codes

JohnNzj committed:
Modify Parser.java: too many text in a sentence

JohnNzj committed:
Command.java: Add find, Task.java: Add getDescription, Ui.java: Add printFind

JohnNzj committed:
Merge branch 'branch-A-JavaDoc'

JohnNzj committed:
Merge branch 'branch-A-CodingStandard'

JohnNzj committed:
Merge branch 'branch-Level-9'

Ellieyee reviewed (commented):
For Parser.java: 
Can consider using a switch case instead of if-else statements as it can be harder to trace the effects. The switch case implemented in line 42 can be abstracted out into a separate function for readability (or can consider storing pre-loading into a hashmap). Overall, comments are concise and accurately follows the javadoc coding standards 👍

Ellieyee commented on src/main/java/Ui.java:
```diff
@@ -124,5 +124,16 @@ public void listTask(TaskList tasks) {
         }
     }
+
+    public void printFind(String keyword) {
```
Can consider adding an alternative message if no matching tasks are found.

Ellieyee commented on src/main/java/Command.java:
```diff
@@ -72,6 +72,10 @@ public void execute(TaskList tasks, Ui ui, Storage storage) {
             tasks.delete(deletionNumber);
             ui.printDelete(toDelete);
             break;
+        case "find":
```
Each case can be extracted out into separate functions or sub-classes for readability.

Ellieyee commented on src/main/java/Deadline.java:
```diff
@@ -1,12 +1,26 @@
+     * Creates a deadline task.
+     *
+     * @param description A string representing the description of the task
+     * @param by A string representing the deadline of that task.
```
String "by" can be renamed to clearer description E.g taskDeadline

Ellieyee commented on src/main/java/Event.java:
```diff
@@ -1,12 +1,25 @@
+     * Creates a Task of type Event.
+     *
+     * @param desription A string representing the task's description.
+     * @param when A string representing time of the task.
```
String when can be renamed to clearer description E.g eventDuration

ang-zeyu commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,100 @@
+     * @param ui An Ui object that helps to interact with the user.
+     * @param storage A Storage object that contains a file with the tasks in it.
+     */
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
```
Since commands are limited to a fixed set of values, maybe it would be a good idea to use an enum?

I think it also offers a trivial runtime benefit as you only call CommandsEnum.valueOf(taskName) once ( and then catch IllegalArgumentException if invalid), and compare for enum equality.
( In C / C++ / C# enums are fundamentally implemented with integers, I believe it is the same for Java but I cant find a good source on it thus far )

(instead of comparing string equality till the correct string is reached)

ang-zeyu replied:
As per Ellieyee's comment below, you could also extract and make abstractions for each of the individual commands.

ang-zeyu commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,46 @@
+    /**
+     * Loads and saves files by applying other classes.
+     */
+    public void run() {
```
👍 Effective and clean REPL!

ang-zeyu commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,28 @@
+ */
+public class Event extends Task {
+
+    protected String when;
```
When could be renamed to a more semantic 'period', or a seperate 'start', and 'end date'!

ang-zeyu commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,120 @@
+     * @param time A string representing date in the form of dd/mm/yyyy hhmm
+     * @return A string representation of a date in english sentences.
+     */
+    public static String convertDateAndTime(String time) { // in the form of dd/mm/yyyy hhmm
```
You could use java's inbuilt java.time (aka JodaTime) library to do the parsing ( and even formatting output ) instead!
https://docs.oracle.com/javase/8/docs/api/java/time/package-summary.html
Classes relavant ( I believe ): DateTimeFormatter, TimeFormatter, LocalDateTime, LocalTime

Regarding the java documentation (while browsing the docs):
There are many date time related classes outside of the java.time which are deprecated and harder to use (pre java8 apis).

FelixNWJ replied:
As what Zeyu stated above, unless your particular output format is not able to be returned by any methods in the java Time classes, its better to use their library since it saves readers the need to read your implementation and makes the code look neater as well.

ang-zeyu commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,201 @@
+            for (int i = 1; i <= TaskList.size(); i++) {
+                if (i == number) {
+                    if (i == TaskList.size()) {
+                        s = s + scanFile.nextLine().replaceFirst("0", "1");
```
What if a task had already been marked as done, and the task description has a 0 in it?

_Suggestion \:_
- Either split the nextLine by the pipeline character seperator '|' and edit the second element of the array
- Or change the done command not to execute if the task was already marked as done!

ang-zeyu commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,201 @@
+            Scanner sc = new Scanner(file);
+            while (sc.hasNext()) {
+                String type = sc.next();
+                switch (type) {
```
- Another good place to consider the use of 👍 enums 👍  
- Could also check for validity of storage data (e.g. if the file is corrupted / of the wrong format ) and throw the appropriate exceptions if the conditions are not met.

ang-zeyu commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,201 @@
+     *
+     * @param taskname A string representing the task description.
+     */
+    public void addTodo(String taskname) {
```
_Suggestion \:_
Generalise the logic when adding a new task to avoid adding methods every time a new type of Task is added.
For example (pseudo code) \:
    `fileWriter.write(String.format("%s | %s | %s", task.getType().toString(), task.isDone(), task.getDescription());`

ang-zeyu commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,139 @@
+/**
+ * The Ui class provides methods to deal with user interactions
+ */
+public class Ui {
```
Clean abstraction of ui overall 👍

ang-zeyu reviewed (commented):
Missing summary message from first PR review :

👍 Clean code overall! 👍 
Could seperate the abstractions for commands and storage in future iterations to increase code reusability.

JohnNzj committed:
Task.java: Add isDone()
Command.java: Modify case "done"

FelixNWJ commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,104 @@
+     * @param ui An Ui object that helps to interact with the user.
+     * @param storage A Storage object that contains a file with the tasks in it.
+     */
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
```
Since the execute method is going to be different for all commands, its better to make this method (and by extension the class) abstract instead of having a switch case, as it makes your execute method have too much functionality. It is better to break it up instead of possible

FelixNWJ commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,46 @@
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
+        while (!isExit) {
```
Systematic coding of each individual action that you want to execute to a corresponding method.

FelixNWJ commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,201 @@
+     *
+     * @param number An int representing the task number.
+     */
+    public void updateDone(int number) {
```
Method looks succint and each line does one task, well done!

FelixNWJ commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * The Task class represents the things to do.
+ */
+public class Task {
```
Since the Task class is primarily used for inheritance for other subclasses, and we do not actually save a Task parent object, its better to make the class abstract instead.

FelixNWJ commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,65 @@
+ */
+public class TaskList {
+
+    private static ArrayList<Task> tasks;
```
In the event that your Duke uses many TaskLists for a certain feature, I think that the tasks list should be implemented as a non static attribute, since one tasklist should not be used for all Duke objects created.

FelixNWJ commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,139 @@
+            if (TaskList.getList().get(i).getDescription().contains(keyword)) {
+                System.out.println(count + "." + TaskList.getList().get(i));
+                count++;
+            }
```
Code looks clean here and focuses on the functionality of the UI, well done!

FelixNWJ commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,120 @@
+     *
+     * @param command A string representing the user's command.
+     * @return A Command object.
+     */
```
Is there a need to create a Parser in this case? Since it directly returns a Command object without much modification

FelixNWJ reviewed (commented):
Overall code is clean and neat, while adhering to the coding style guide. However, abstractions and modularisation could be implemented more so that the code is more readable

JohnNzj committed:
Remove static from methods

JohnNzj committed:
Merge branch 'gradle'

JohnNzj committed:
build.gradle: setup gradle to run duke and unit tests

JohnNzj committed:
Modify gradle for CheckStyle

JohnNzj committed:
Add MainWindow.fxml, DialogBox.fxml, image files, Main.java, MainWindow.java, Launcher.java,
DialogBox.java
Ui.java: Modify void to return String
Duke.java: Modify duke for GUI
Command.java Modify execute to return String

JohnNzj committed:
Merge branch 'branch-A-Gradle'

JohnNzj committed:
Merge branch 'branch-A-CheckStyle'

JohnNzj committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	build.gradle
#	src/main/java/Duke.java
#	src/main/java/Ui.java

JohnNzj committed:
duke.txt: Modified tasks
Ui.java: JavaDocs
MainWindow.java: Implemented "bye" & welcome text
Duke.java: JavaDocs
DialogBox.java: JavaDocs
Command.java: Indented properly

JohnNzj committed:
build.gradle: include shadowJar

JohnNzj committed:
Handle few more exceptions

JohnNzj committed:
Command.java: Add assertions as various points

JohnNzj committed:
Merge pull request #5 from JohnNzj/branch-A-Assertions

Command.java: Add assertions as various points

JohnNzj committed:
Merge branch 'master' of https://github.com/JohnNzj/duke

JohnNzj committed:
Command.java: Add guard clauses (#6)

Storage.java: Combine else if into one to prevent reuse

JohnNzj committed:
Merge branch 'master' of https://github.com/JohnNzj/duke

JohnNzj committed:
Add DoAfter.java
Modify: Parser.java
Improve: Storage.java
Modify Command.java for DoAfter task

JohnNzj committed:
Storage.java: Abstract add different task into 1
Command.java: Modify method parameters

JohnNzj committed:
Storage.java: Add method to create file if it does not exist

JohnNzj committed:
Add Ui.png

JohnNzj committed:
Add Ui.png

JohnNzj committed:
Revert "Add Ui.png"

This reverts commit 36496fb63c8caae57560bbc0fa4e358096a152fb.

JohnNzj committed:
Set theme jekyll-theme-hacker

JohnNzj committed:
Update README.md

JohnNzj committed:
Update README.md

JohnNzj committed:
Set theme jekyll-theme-hacker

JohnNzj committed:
Update README.md

JohnNzj committed:
Update README.md

JohnNzj committed:
Merge branch 'master' of https://github.com/JohnNzj/duke

JohnNzj committed:
build.gradle: modify for cross platforms

JohnNzj committed:
MainWindow.java: Add setPrefHeight to prevent truncation
